### PR TITLE
Add retry support when GD endpoint is unvailable

### PR DESCRIPTION
**Pull request by sadam21**
Currently only for GET method and some kinds of HTTP 5XX status.

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: src/main/java/com/gooddata/GoodDataSettings.java [hunk 1]
```diff
@@ -205,30 +208,32 @@ public void setUserAgent(String userAgent) {
...
         final GoodDataSettings that = (GoodDataSettings) o;
-
-        if (maxConnections != that.maxConnections) return false;
-        if (connectionTimeout != that.connectionTimeout) return false;
-        if (connectionRequestTimeout != that.connectionRequestTimeout) return false;
-        if (socketTimeout != that.socketTimeout) return false;
-        if (pollSleep != that.pollSleep) return false;
-        return userAgent != null ? userAgent.equals(that.userAgent) : that.userAgent == null;
+        return maxConnections == that.maxConnections &&
+                connectionTimeout == that.connectionTimeout &&
```
Comment (jimirocks) [inline src/main/java/com/gooddata/GoodDataSettings.java]: consider to put `&&` at the beginning

File: src/main/java/com/gooddata/retry/FailOverSettings.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2007-2017, GoodData(R) Corporation. All rights reserved.
```
Comment (jimirocks) [inline src/main/java/com/gooddata/retry/FailOverSettings.java]: shouldn't we update year in the template?

File: src/main/java/com/gooddata/retry/FailOverSettings.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Total retry count. Should be > 1. No retry if not set.
+     * @return retry count
+     */
+    public Integer getRetryCount() {
+        return retryCount;
+    }
+
+    public void setRetryCount(Integer retryCount) {
+        this.retryCount = retryCount;
```
Comment (jimirocks) [inline src/main/java/com/gooddata/retry/FailOverSettings.java]: shouldn't it check the validity?

File: src/main/java/com/gooddata/retry/RetryableRestTemplate.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        notNull(retryTemplate);
+        this.retryTemplate = retryTemplate;
+        this.failOverStrategy = failOverStrategy;
+    }
+
+    @Override
+    protected <T> T doExecute(URI url, HttpMethod method, RequestCallback requestCallback,
+                              ResponseExtractor<T> responseExtractor) throws RestClientException {
+        return retryTemplate.execute(context -> {
+            System.out.println("Retring " + context.getRetryCount() + " " + method + " ");
```
Comment (jimirocks) [inline src/main/java/com/gooddata/retry/RetryableRestTemplate.java]: Typo...

File: src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        "DELETE" | 500                 | false            | GoodDataException     | MIN_NO_RETRY_DURATION | MAX_NO_RETRY_DURATION
+        "DELETE" | 501                 | false            | GoodDataException     | MIN_NO_RETRY_DURATION | MAX_NO_RETRY_DURATION
+        "PUT"    | 500                 | false            | GoodDataRestException | MIN_NO_RETRY_DURATION | MAX_NO_RETRY_DURATION
+        "PUT"    | 501                 | false            | GoodDataRestException | MIN_NO_RETRY_DURATION | MAX_NO_RETRY_DURATION
+    }
+
+    void cleanup() {
+        closeJadler()
+    }
+
```
Comment (jimirocks) [inline src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy]: too many blank lines

File: src/main/java/com/gooddata/retry/RetryableRestTemplate.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.web.client.RestTemplate;
+
+import java.net.URI;
+
+import static org.apache.commons.lang3.Validate.notNull;
+
+/**
+ * REST template with retry ability. Its behavior is described by given strategy and retry template.
+ */
+public class RetryableRestTemplate extends RestTemplate {
```
Comment (jimirocks) [inline src/main/java/com/gooddata/retry/RetryableRestTemplate.java]: Hmm it's kind of weird mixin Retryable and FailOver....let's discuss in person

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: src/main/java/com/gooddata/GoodDataSettings.java (hunk 1 above)
Comment (sadam21) [reply src/main/java/com/gooddata/GoodDataSettings.java]: Done

File: src/main/java/com/gooddata/retry/FailOverSettings.java (hunk 2 above)
Comment (sadam21) [reply src/main/java/com/gooddata/retry/FailOverSettings.java]: Fixed

File: src/main/java/com/gooddata/retry/FailOverSettings.java (hunk 3 above)
Comment (sadam21) [reply src/main/java/com/gooddata/retry/FailOverSettings.java]: Added

File: src/main/java/com/gooddata/retry/RetryableRestTemplate.java (hunk 4 above)
Comment (sadam21) [reply src/main/java/com/gooddata/retry/RetryableRestTemplate.java]: Fixed, thanks

File: src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy (hunk 5 above)
Comment (sadam21) [reply src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy]: Removed

File: src/main/java/com/gooddata/retry/RetryableRestTemplate.java (hunk 6 above)
Comment (sadam21) [reply src/main/java/com/gooddata/retry/RetryableRestTemplate.java]: ok :-)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: pom.xml [hunk 7]
```diff
@@ -311,6 +311,12 @@
             <version>${spring.version}</version>
             <optional>true</optional>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.retry</groupId>
+            <artifactId>spring-retry</artifactId>
+            <version>1.2.4.RELEASE</version>
+            <optional>true</optional>
```
Comment (jimirocks) [inline pom.xml]: should we somewhere document, one needs to add `spring-retry` when configuring retries?

File: src/main/java/com/gooddata/GoodData.java [hunk 8]
```diff
@@ -240,7 +246,7 @@ protected GoodData(GoodDataEndpoint endpoint, Authentication authentication, Goo
         lcmService = new LcmService(getRestTemplate(), settings);
     }
 
-    static RestTemplate createRestTemplate(GoodDataEndpoint endpoint, HttpClient httpClient) {
+    static RestTemplate createRestTemplate(GoodDataEndpoint endpoint, HttpClient httpClient, RetrySettings failOverSettings) {
```
Comment (jimirocks) [inline src/main/java/com/gooddata/GoodData.java]: please rename as well

File: src/main/java/com/gooddata/GoodData.java [hunk 9]
```diff
@@ -263,6 +274,30 @@ static RestTemplate createRestTemplate(GoodDataEndpoint endpoint, HttpClient htt
         return restTemplate;
     }
 
+    private static RestTemplate createFailOverRestTemplate(RetrySettings failOverSettings, UriPrefixingClientHttpRequestFactory factory) {
```
Comment (jimirocks) [inline src/main/java/com/gooddata/GoodData.java]: forgotten rename

File: src/main/java/com/gooddata/retry/RetryableRestTemplate.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    private final RetryTemplate retryTemplate;
+    private final RetryStrategy failOverStrategy;
+
+    /**
+     * Create a new instance of the {@link RetryableRestTemplate}.
+     * @param requestFactory HTTP request factory to use
+     * @param retryTemplate retry template
+     * @param failOverStrategy failover strategy
```
Comment (jimirocks) [inline src/main/java/com/gooddata/retry/RetryableRestTemplate.java]: rename

File: src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    private static final long MIN_RETRY_DURATION = 3000l
+
+    @Override
+    protected ConnectorService getService() {
+        return gd.getConnectorService()
+    }
+
+    @Override
+    protected GoodDataSettings createGoodDataSettings() {
+        def failOverSettings = new RetrySettings()
```
Comment (jimirocks) [inline src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy]: rename

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: README.md [hunk 12]
```diff
@@ -37,6 +37,19 @@ The *GoodData Java SDK* uses:
 * the *Slf4j API* version 1.7.*
 * the *Java Development Kit (JDK)* version 8 or later
 
+##### Retry of failed API calls
+
+Since version *2.33.1* you can retry your failed requests. Turn it on by configuring
+[RetrySettings](https://github.com/sadam21/gooddata-java/blob/ast-failover_retry/src/main/java/com/gooddata/retry/RetrySettings.java)
```
Comment (jimirocks) [inline README.md]: please refer to origin/master version

File: src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy [hunk 13]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2007-2017, GoodData(R) Corporation. All rights reserved.
```
Comment (liry) [inline src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy]: Still old year in some copyrights :)

File: README.md [hunk 14]
```diff
@@ -37,6 +37,19 @@ The *GoodData Java SDK* uses:
...
+and add [Spring retry](https://github.com/spring-projects/spring-retry) to your classpath:
+```xml
+        <dependency>
+            <groupId>org.springframework.retry</groupId>
+            <artifactId>spring-retry</artifactId>
+            <version>1.2.4.RELEASE</version>
```
Comment (liry) [inline README.md]: Is hard coded version in doc good idea?

File: README.md [hunk 15]
```diff
@@ -37,6 +37,19 @@ The *GoodData Java SDK* uses:
 * the *Slf4j API* version 1.7.*
 * the *Java Development Kit (JDK)* version 8 or later
 
+##### Retry of failed API calls
+
+Since version *2.33.1* you can retry your failed requests. Turn it on by configuring
```
Comment (liry) [inline README.md]: What about to raise minor version with this new feature? It's not a bug fix ;)

File: README.md (hunk 12 above)
Comment (liry) [inline README.md]: Please replace your fork with gooddata ;)

File: pom.xml (hunk 7 above)
Comment (liry) [reply pom.xml]: I would also add comment here, that you need this dependency when using `RetrySettings` ;)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: README.md (hunk 12 above)
Comment (sadam21) [reply README.md]: Done

File: README.md (hunk 15 above)
Comment (sadam21) [reply README.md]: OK :-)

File: README.md (hunk 14 above)
Comment (sadam21) [reply README.md]: Not sure, replaced by property.

File: src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy (hunk 13 above)
Comment (sadam21) [reply src/test/groovy/com/gooddata/retry/RetryableRestTemplateTest.groovy]: Fixed.

File: README.md (hunk 12 above)
Comment (sadam21) [reply README.md]: Done.

**Comment by jimirocks** [comment]
OK to me, but pls remove `.attach_pid206` from changeset

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: src/main/java/com/gooddata/retry/RetrySettings.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    private Long DEFAULT_RETRY_MAX_INTERVAL = 1 * 60 * 1000l; // 1min
+    private Double DEFAULT_RETRY_MULTIPLIER = 2d;
+
+    private Integer retryCount = DEFAULT_RETRY_COUNT;
+    private Long retryInitialInterval = DEFAULT_RETRY_INITIAL_INTERVAL;
+    private Long retryMaxInterval = DEFAULT_RETRY_MAX_INTERVAL;
+    private Double retryMultiplier = DEFAULT_RETRY_MULTIPLIER;
+
+    /**
+     * Total retry count. Should be > 1. No retry if not set.
```
Comment (liry) [inline src/main/java/com/gooddata/retry/RetrySettings.java]: Really > 1? Setter does check only > 0

File: pom.xml (hunk 7 above)
Comment (sadam21) [reply pom.xml]: Added.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: src/main/java/com/gooddata/retry/RetrySettings.java (hunk 16 above)
Comment (sadam21) [reply src/main/java/com/gooddata/retry/RetrySettings.java]: Corrected :-)

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-java-merge-compliance-check/93/ : SUCCESS in 29s

**Comment by liry** [comment]
Resolves #782.

**Pull request merged**